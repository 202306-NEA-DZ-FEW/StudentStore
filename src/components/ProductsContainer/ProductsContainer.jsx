import { useEffect, useRef, useState } from "react";
import Filter from "../Filter/Filter";
import ProductsList from "../ProductsList/ProductsList";

export default function ProductsContainer({ products }) {
    const [filteredProducts, setFilteredProducts] = useState([]);
    const [minPrice, setMinPrice] = useState("");
    const [maxPrice, setMaxPrice] = useState("");
    const [sortByPriceAsc, setSortByPriceAsc] = useState(false);
    const [sortByPriceDesc, setSortByPriceDesc] = useState(false);
    // debounce functionality
    const debounceTimeout = useRef(null);
    const debounce = (func, delay) => {
        return function (...args) {
            if (debounceTimeout.current) {
                clearTimeout(debounceTimeout.current);
            }

            debounceTimeout.current = setTimeout(() => {
                func(...args);
            }, delay);
        };
    };
    //  min price change function
    const handleMinPriceChange = (e) => {
        const newMinPrice = e.target.value;
        setMinPrice(newMinPrice);
        debounce(filterProducts, 500)(newMinPrice, maxPrice);
    };
    // max price change function
    const handleMaxPriceChange = (e) => {
        const newMaxPrice = e.target.value;
        setMaxPrice(newMaxPrice);
        debounce(filterProducts, 500)(minPrice, newMaxPrice);
    };
    // filter price functions
    const filterProducts = (min, max) => {
        const filtered = products.filter((product) => {
            const price = parseFloat(product.price);
            const minPriceValue = parseFloat(min);
            const maxPriceValue = parseFloat(max);
            return (
                isNaN(price) ||
                ((isNaN(minPriceValue) || price >= minPriceValue) &&
                    (isNaN(maxPriceValue) || price <= maxPriceValue))
            );
        });

        setFilteredProducts(filtered);
    };
    // sorting by asc price
    const handleSortByPriceAscChange = () => {
        setSortByPriceAsc(!sortByPriceAsc);
        setSortByPriceDesc(false);
    };
    // sorting by des price
    const handleSortByPriceDescChange = () => {
        setSortByPriceDesc(!sortByPriceDesc);
        setSortByPriceAsc(false);
    };
    let sortedProducts = [...filteredProducts];

    if (sortByPriceAsc) {
        sortedProducts.sort(
            (a, b) => parseFloat(a.price) - parseFloat(b.price)
        );
    } else if (sortByPriceDesc) {
        sortedProducts.sort(
            (a, b) => parseFloat(b.price) - parseFloat(a.price)
        );
    }
    // using this useEffect to diplay all product at the firt render
    useEffect(() => {
        setFilteredProducts(products);
    }, [products]);
    return (
        <div className='flex'>
            <Filter
                minPrice={minPrice}
                maxPrice={maxPrice}
                handleMinPriceChange={handleMinPriceChange}
                handleMaxPriceChange={handleMaxPriceChange}
                sortByPriceAsc={sortByPriceAsc}
                sortByPriceDesc={sortByPriceDesc}
                handleSortByPriceAscChange={handleSortByPriceAscChange}
                handleSortByPriceDescChange={handleSortByPriceDescChange}
            />
            <ProductsList products={sortedProducts} />
        </div>
    );
}
