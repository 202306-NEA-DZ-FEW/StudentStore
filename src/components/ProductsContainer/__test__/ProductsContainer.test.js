import renderer from "react-test-renderer";
import ProductsContainer from "../ProductsContainer";

jest.mock("next/router", () => ({
    useRouter: () => ({
        pathname: "/",
        locale: "en",
    }),
}));
const t = (key) => key;
it("renders correctly", () => {
    const tree = renderer.create(<ProductsContainer t={t} />).toJSON();
    expect(tree).toMatchSnapshot();
});
