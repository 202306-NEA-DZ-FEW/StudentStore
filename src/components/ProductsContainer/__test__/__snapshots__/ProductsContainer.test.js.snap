// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="flex"
>
  <div
    className="w-1/4"
  >
    <div>
      <h2>
        Price
      </h2>
      <div
        className="ml-4"
      >
        <div
          className="flex space-x-2 justify-center my-2"
        >
          <input
            className="w-1/2 h-6 border rounded-md px-2 focus:outline-none"
            onChange={[Function]}
            placeholder="Min"
            type="number"
            value=""
          />
          <input
            className="w-1/2 h-6 border rounded-md px-2 focus:outline-none"
            onChange={[Function]}
            placeholder="Max"
            type="number"
            value=""
          />
        </div>
      </div>
    </div>
    <h2>
      Sort
    </h2>
    <div
      className="ml-4"
    >
      <div
        className="flex items-center mb-1"
      >
        <input
          checked={false}
          className="mr-2"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          className="checkbox-label"
        >
          Price Low to High
        </label>
      </div>
      <div
        className="flex items-center mb-1"
      >
        <input
          checked={false}
          className="mr-2"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          className="checkbox-label"
        >
          Price High to Low
        </label>
      </div>
    </div>
    <div>
      <h2>
        Type
      </h2>
      <div
        className="ml-4"
      >
        <div
          className="mb-1"
        >
          <input
            checked={false}
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="label-title ml-2 mb-1 capitalize"
          >
            for sale
          </label>
        </div>
        <div
          className="mb-1"
        >
          <input
            checked={false}
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="label-title ml-2 mb-1 capitalize"
          >
            for borrow
          </label>
        </div>
      </div>
    </div>
    <div>
      <h2>
        Category
      </h2>
      <div
        className="ml-4"
      >
        <div
          className="mb-1"
        >
          <input
            checked={false}
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="label-title ml-2 mb-1 capitalize"
          >
            electronics
          </label>
        </div>
        <div
          className="mb-1"
        >
          <input
            checked={false}
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="label-title ml-2 mb-1 capitalize"
          >
            books
          </label>
        </div>
        <div
          className="mb-1"
        >
          <input
            checked={false}
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="label-title ml-2 mb-1 capitalize"
          >
            gaming
          </label>
        </div>
        <div
          className="mb-1"
        >
          <input
            checked={false}
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="label-title ml-2 mb-1 capitalize"
          >
            clothes
          </label>
        </div>
        <div
          className="mb-1"
        >
          <input
            checked={false}
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="label-title ml-2 mb-1 capitalize"
          >
            shoes
          </label>
        </div>
        <div
          className="mb-1"
        >
          <input
            checked={false}
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="label-title ml-2 mb-1 capitalize"
          >
            food
          </label>
        </div>
        <div
          className="mb-1"
        >
          <input
            checked={false}
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="label-title ml-2 mb-1 capitalize"
          >
            transportation
          </label>
        </div>
        <div
          className="mb-1"
        >
          <input
            checked={false}
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="label-title ml-2 mb-1 capitalize"
          >
            furniture
          </label>
        </div>
      </div>
    </div>
  </div>
  <div
    className="w-3/4 flex justify-center flex-wrap gap-4"
  />
</div>
`;
